src/bpf/bpfd-manager.c

```diff
@@ -0,0 +1,489 @@
+                                        JSON_BUILD_PAIR_UNSIGNED("type", prog_info.type),
+                                        JSON_BUILD_PAIR_UNSIGNED("memlock", memlock),
+                                        JSON_BUILD_PAIR_STRING("tag", prog_tag),
+                                        JSON_BUILD_PAIR_CONDITION(prog_map_names, "map_names", JSON_BUILD_STRV(prog_map_names)),
+                                        JSON_BUILD_PAIR_UNSIGNED("run_time_ns", prog_info.run_time_ns),
+                                        JSON_BUILD_PAIR_UNSIGNED("run_count", prog_info.run_cnt))));
+                if (r < 0) {
+                        log_error_errno(r, "Failed to build JSON object for program, ignoring: %m");
+                        continue;
+                }
+
+                if (link) {
+                        /* Check if this is the last bpf program. If it is, varlink_reply instead of varlink_notify */
+                        if (sym_bpf_prog_get_next_id(id, &next_id) >= 0) {
+                                r = varlink_notify(link, v);
+                                if (r < 0) {
+                                        log_error_errno(r, "Failed to varlink notify for program, ignoring: %m");
+                                        continue;
+                                }
+                        } else {
+                                r = varlink_reply(link, v);
+                                if (r < 0) {
+                                        log_error_errno(r, "Failed to varlink reply for program, ignoring: %m");
+                                        continue;
+                                }
+                        }
+                } else {
+                        r = json_variant_append_array(&prog_and_map_list, v);
+                        if (r < 0) {
+                                log_error_errno(r, "Failed to append to json variant array for program, ignoring: %m");
+                                continue;
+                        }
+                        /* If this is the last bpf program, log everything we've accumulated */
+                        if (sym_bpf_prog_get_next_id(id, &next_id) < 0) {
+                                _cleanup_(erase_and_freep) char *formatted = NULL;
+
+                                r = json_variant_format(prog_and_map_list, JSON_FORMAT_NEWLINE, &formatted);
+                                if (r < 0) {
+                                        log_error_errno(r, "Failed to format list of bpf programs and maps, ignoring: %m");
+                                        continue;
+                                }
+
+                                log_struct(LOG_INFO, "MESSAGE_ID=" SD_MESSAGE_BPFD_LOG_STR, "DATA=%s", formatted);
+                        }
+                }
+        }
+
+        char *map_name;
+        HASHMAP_FOREACH(map_name, map_name_by_id)
+                free(map_name);
```

Comment:
please use the hash_ops stuff so that the when the hashmap is freed it automatically frees its remaining entries.

Reply:
I used _cleanup_hashmap_free_free_ instead of hash_ops, since I couldn't find one that frees the values only. This hashmap simply maps from a uint32_t to a string, so I think that should be sufficient.

Reply:
Also, if you wouldn't mind double checking that I'm freeing the Hashmap and JsonVariant stuff properly, I would really appreciate it. I'm not confident whether or not I'm doing it correctly, especially with passing JsonVariant pointers around.